trace-dispatcher/src/Cardano/Logging/Tracer/Standard.hs

```diff
@@ -0,0 +1,94 @@
+import           Data.IORef (IORef, modifyIORef, newIORef, readIORef)
+import           Data.Text (Text)
+import qualified Data.Text.IO as TIO
+import           GHC.Conc (ThreadId)
+import           System.IO (hFlush, stdout)
+
+import           Cardano.Logging.DocuGenerator
+import           Cardano.Logging.Types
+
+import qualified Control.Tracer as T
+
+-- | Do we log to stdout or to a file?
+data LogTarget = LogStdout | LogFile FilePath
+  deriving (Eq, Show)
+
+-- | The state of a standard tracer
+data StandardTracerState a =  StandardTracerState {
+    stRunning :: Maybe (InChan Text, OutChan Text, ThreadId)
+  , stTarget  :: LogTarget
+}
+
+emptyStandardTracerState :: Maybe FilePath -> StandardTracerState a
+emptyStandardTracerState Nothing   = StandardTracerState Nothing LogStdout
+emptyStandardTracerState (Just fp) = StandardTracerState Nothing (LogFile fp)
+
+
+standardTracer :: forall m. (MonadIO m)
+  => Maybe FilePath
+  -> m (Trace m FormattedMessage)
+standardTracer mbFilePath = do
+    stateRef <- liftIO $ newIORef (emptyStandardTracerState mbFilePath)
+    pure $ Trace $ T.arrow $ T.emit $ uncurry3 (output stateRef)
+  where
+    output ::
+         IORef (StandardTracerState a)
+      -> LoggingContext
+      -> Maybe TraceControl
+      -> FormattedMessage
+      -> m ()
+    output stateRef LoggingContext {} Nothing (FormattedHuman _c msg) = liftIO $ do
+      st  <- readIORef stateRef
+      case stRunning st of
+        Just (inChannel, _, _) -> writeChan inChannel msg
+        Nothing                -> pure ()
+    output stateRef LoggingContext {} Nothing (FormattedMachine msg) = liftIO $ do
+      st  <- readIORef stateRef
+      case stRunning st of
+        Just (inChannel, _, _) -> writeChan inChannel msg
+        Nothing                -> pure ()
+    output stateRef LoggingContext {} (Just Reset) _msg = liftIO $ do
+      st <- readIORef stateRef
+      case stRunning st of
+        Nothing -> initLogging stateRef
+        Just _  -> pure ()
+    output _ lk (Just c@Document {}) (FormattedHuman co msg) =
+       docIt
+        (Stdout (if co then HumanFormatColoured else HumanFormatUncoloured))
+        (FormattedHuman co "")
+        (lk, Just c, msg)
+    output _ lk (Just c@Document {}) (FormattedMachine msg) =
+       docIt (Stdout MachineFormat) (FormattedMachine "") (lk, Just c, msg)
+    output _stateRef LoggingContext {} _ _a = pure ()
+
+-- TODO: care about reconfiguration
+initLogging :: IORef (StandardTracerState a) -> IO ()
+initLogging stateRef = do
+  (inChan, outChan) <- newChan 2048
+  threadId <- forkIO $ forever $ do
+    msg   <- readChan outChan
+    state <- readIORef stateRef
+    case stTarget state of
+        LogFile f -> do
+                        TIO.appendFile f msg
+                        TIO.appendFile f "\n"
+        LogStdout -> do
+                        TIO.putStrLn msg
+                        hFlush stdout
+  modifyIORef stateRef (\ st ->
```

Comment:
Sounds scary. But what kind of async exception should happen here?

Reply:
User hits `Ctrl-c`, lack of memory, or what more common killing a thread with `throwTo` or `cancel`. See [AsyncException](https://hackage.haskell.org/package/base-4.15.0.0/docs/Control-Exception-Base.html#t:AsyncException).

Reply:
@coot What code change would you suggest?

Reply:
I'd use `withAsync` - the caller will need to make sure that it does not exit from its continuation too early (which would kill the logging thread), and it also has a chance to do something if the logging thread dies.  since you use `forever` the async can have a type `Async Void` which indicates that the spawn thread will never produce a value, it could only throw an exception.

Reply:
I changed to async, but in my version of withAsync, the channel read blocks forever, no idea why? I checked the not working code in as comment. Can you see what I'm doing wrong?

Reply:
Have you pushed it? I don't see the non working version.  If you try to use the channel outside of the continuation it would block, because the thread would be dead.  Could that be the cause?

Reply:
@coot should be their: 
    -- this doesn't work for unknown reason
    -- withAsync (writerThread stateRef outChan) (\ as ->
    --   modifyIORef stateRef (\ st ->
    --     st {stRunning = Just (inChan, outChan, as)})

Reply:
1. We have dropped the file writer option
2. We consider `stdout` writes blocking to be an extreme corner case -- since it's only possible with `stdout` redirection to a file system -- and also considering that no reasonable deployment would do this, since it makes log rotation impossible.